ansible_collections/arista/avd/roles/eos_designs/python_modules/network_services/router_bgp.py

```diff
@@ -70,34 +73,33 @@ def _router_bgp_peer_groups(self) -> list | None:
 
         peer_groups = []
         peer_peergroups = set()
+        af_peer_groups = {}
         for tenant in self.shared_utils.filtered_tenants:
             for vrf in tenant["vrfs"]:
                 # bgp_peers is already filtered in filtered_tenants to only contain entries with our hostname
                 if not (vrf["bgp_peers"] or vrf.get("bgp_peer_groups")):
                     continue
 
                 vrf_peer_peergroups = set([peer["peer_group"] for peer in vrf["bgp_peers"] if "peer_group" in peer])
-                peer_groups.extend(
-                    [
-                        peer_group
-                        for peer_group in vrf.get("bgp_peer_groups", [])
-                        if (self.shared_utils.hostname in peer_group.get("nodes", []) or peer_group["name"] in vrf_peer_peergroups)
-                    ]
-                )
+                for peer_group in vrf.get("bgp_peer_groups", []):
+                    if self.shared_utils.hostname in peer_group.get("nodes", []) or peer_group["name"] in vrf_peer_peergroups:
+                        peer_groups.append(peer_group)
+                        self._address_family_peer_groups(peer_group["name"], af_peer_groups, peer_group, "address_family_ipv4")
+                        self._address_family_peer_groups(peer_group["name"], af_peer_groups, peer_group, "address_family_ipv6")
                 peer_peergroups.update(vrf_peer_peergroups)
 
-            peer_groups.extend(
-                [
-                    peer_group
-                    for peer_group in tenant.get("bgp_peer_groups", [])
-                    if (self.shared_utils.hostname in peer_group.get("nodes", []) or peer_group["name"] in peer_peergroups)
-                ]
-            )
+            for peer_group in tenant.get("bgp_peer_groups", []):
+                if self.shared_utils.hostname in peer_group.get("nodes", []) or peer_group["name"] in peer_peergroups:
+                    peer_groups.append(peer_group)
+                    self._address_family_peer_groups(peer_group["name"], af_peer_groups, peer_group, "address_family_ipv4")
+                    self._address_family_peer_groups(peer_group["name"], af_peer_groups, peer_group, "address_family_ipv6")
 
         bgp_peer_groups = []
         if peer_groups:
```

Comment:
I think it would be better to implement the per-address family handling together with the rest. Put it in the helper function and only call it once. And keep everything above as it was.
It will make it simpler also to build the complete result dict and just append to each of the lists as needed.

Reply:
Made the changes.